File: leetcode/107. Binary Tree Level Order Traversal II.js
#!/usr/bin/env node

'use strict';

//https://leetcode.com/problems/binary-tree-level-order-traversal-ii/description/

// Given a binary tree, return the bottom-up level order traversal of its nodes' values. (ie, from left to right, level by level from leaf to root).
//
// For example:
//     Given binary tree [3,9,20,null,null,15,7],
// 3
// / \
// 9  20
// /  \
// 15   7
// return its bottom-up level order traversal as:
// [
//     [15,7],
//     [9,20],
//     [3]
// ]
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[][]}
 */
var levelOrderBottom = function (root) {
    let wrapList = [];
    levelMaker(wrapList, root, 0);
    return wrapList;

    function levelMaker(list, root, level) {
        if (!root) return;
        if (level >= list.length) {
            list.unshift([]);
        }
        levelMaker(list, root.left, level + 1);
        levelMaker(list, root.right, level + 1);
        list[list.length - level - 1].push(root.val);
    }
};

// BFS
var levelOrderBottom2 = function (root) {
    let res = [];
    if (!root) {
        return res;
    }
    let queue = [];
    queue.push(root);
    while (queue.length) {
        let levelList = [];
        let queueLen = queue.length;
        for (let i = 0; i < queueLen; ++i) {
            const cur = queue.shift();
            levelList.push(cur.val);
            if (cur.left)
                queue.push(cur.left);
            if (cur.right)
                queue.push(cur.right);
        }
        //do not use insert() here,it cost too much time.
        //use reverse() insteadly
        res.push(levelList);
    }
    return res.reverse();
};

File: leetcode/107. Binary Tree Level Order Traversal II.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

# Given a binary tree, return the bottom-up level order traversal of its nodes' values. (ie, from left to right, level by level from leaf to root).
#
# For example:
# Given binary tree [3,9,20,null,null,15,7],
#     3
#    / \
#   9  20
#     /  \
#    15   7
# return its bottom-up level order traversal as:
# [
#   [15,7],
#   [9,20],
#   [3]
# ]

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Solution:
    def levelOrderBottom(self, root):
        """
        :type root: TreeNode
        :rtype: List[List[int]]
        """
        
        def level_maker(list, root, level):
            if not root:
                return
            if level >= len(list):
                list.insert(0, [])
            level_maker(list, root.left, level + 1)
            level_maker(list, root.right, level + 1)
            list[-level - 1].append(root.val)
        
        wrap_list = []
        level_maker(wrap_list, root, 0)
        
        return wrap_list
    
    def levelOrderBottom2(self, root):
        """
        :type root: TreeNode
        :rtype: List[List[int]]
        """
        result = []
        if not root:
            return result
        
        queue = list()
        queue.append(root)
        while len(queue) > 0:
            level_list = []
            for _ in range(len(queue)):
                cur = queue.pop(0)
                level_list.append(cur.val)
                if cur.left:
                    queue.append(cur.left)
                if cur.right:
                    queue.append(cur.right)
            result.append(level_list)
        
        result.reverse()
        return result
